### DISCUSSION TITLE
Help on camera render order and render bins

### DISCUSSION THREAD

**Discussion by Hugal31** [openscenegraph/OpenSceneGraph#1180] [Q&A] [answered]
I am working on a render-to-texture magnifying glass and stumbled on some code playing with the camera render order and render bins. However I didn't find some good documentation on it. I have a few question about it:

* How does the camera render order interacts with the render bins?
* What is the difference between NESTED_RENDER and PRE_RENDER/POST_RENDER and what is it used for?
* Is there a difference between:
  * A setup with a PRE_RENDER camera and a POST_RENDER camera
  * The same setup but the two cameras are POST_RENDER and have a difference render order num.

More generally, where can I find this kind of documentation? I found nothing on the website knowledge base and the OpenSceneGraph cookbook doesn't give a lot of details.

I am using OSG 3.6.4.

Thanks for your help.

**Comment by robertosfield** [comment]
For render-to-texture you want the render the subgraph for it before the main scene, which is set by PRE_RENDER.  Have a look through the examples that do render to texture and you'll see a setting for PRE_RENDER.

For instance these are the examples that use PRE_RENDER (it's used for me than just render-to-texture.):

```
OpenSceneGraph/examples$ grep -r PRE_RENDER .
./osgdistortion/osgdistortion.cpp:        camera->setRenderOrder(osg::Camera::PRE_RENDER);
./osgstereomatch/StereoMultipass.cpp:    _Camera->setRenderOrder(osg::Camera::PRE_RENDER);
./osgstereomatch/StereoMultipass.cpp:    _Camera->setRenderOrder(osg::Camera::PRE_RENDER);
./osgstereomatch/StereoMultipass.cpp:    _Camera->setRenderOrder(osg::Camera::PRE_RENDER);
./osgstereomatch/StereoMultipass.cpp:    _Camera->setRenderOrder(osg::Camera::PRE_RENDER);
./osgstereomatch/StereoPass.cpp:    _Camera->setRenderOrder(osg::Camera::PRE_RENDER);
./osgdepthpeeling/DePee.cpp:  _dePeePasses.back()->Cameras[EDGE_MAP]->setRenderOrder(osg::Camera::PRE_RENDER);
./osgdepthpeeling/DePee.cpp:  pass->Cameras[mapMode]->setRenderOrder(osg::Camera::PRE_RENDER);
./osgposter/osgposter.cpp:    camera->setRenderOrder( osg::Camera::PRE_RENDER );
./osgfpdepth/osgfpdepth.cpp:    rttCamera->setRenderOrder(Camera::PRE_RENDER);
./osgpackeddepthstencil/osgpackeddepthstencil.cpp:    rttCamera->setRenderOrder(osg::Camera::PRE_RENDER);
./osgSSBO/osgSSBO.cpp:        prerenderCamera->setRenderOrder(osg::Camera::PRE_RENDER);
./osgmultiplerendertargets/osgmultiplerendertargets.cpp:        camera->setRenderOrder(osg::Camera::PRE_RENDER);
./osgprerendercubemap/osgprerendercubemap.cpp:        camera->setRenderOrder(osg::Camera::PRE_RENDER);
./osgviewerCocoa/ViewerCocoa.mm:    the_camera->setRenderOrder(osg::Camera::PRE_RENDER);
./osggameoflife/GameOfLifePass.cpp:    _Camera->setRenderOrder(osg::Camera::PRE_RENDER);
./osgprerender/osgprerender.cpp:        camera->setRenderOrder(osg::Camera::PRE_RENDER);
./osgblenddrawbuffers/osgblenddrawbuffers.cpp:    camera->setRenderOrder( osg::Camera::PRE_RENDER );
./osgdeferred/osgdeferred.cpp:    camera->setRenderOrder(osg::Camera::PRE_RENDER);

```

**Reply by Hugal31** [reply]
Thanks, but that doesn't really answer my questions. I already guessed that PRE_RENDER cameras are rendered before the POST_RENDER cameras, but that doesn't tell me how it interacts with render bins, what is NESTED_RENDER is for, etc...

**Reply by robertosfield** [reply]
I'm afraid I don't have time to do a deep dive. I've written explanations about this functionality on the old [osg-users mailing list/forum](https://groups.google.com/g/osg-users) so having a search for NESTED_RENDER would be worth doing a search on.

**Comment by Hugal31** [comment, marked as answer]
After some trial and error, I think I understand the algorithms now. It looks like this:

- For each active camera of the current `GraphicsContext` (which usually are the main and slave camera of the current `Viewer`), sorted by RenderOrder and RenderOrder num: 
  - Traverse the scene graph and "push" the encountered nodes for rendering.
  - If a "nested" (i.e. in the scene graph) camera is encountered:
    - If it has the NESTED_RENDER render order, continue to traverse with its settings (I didn't use this one, so I can't tell for sure).
    - Otherwise, push it in the RenderStage before or after the current camera render, depending on its render order.
    - If I understand correctly, the "nesting" camera do not render the child of the nested cameras.

Each camera does a "render", which includes all the render bins of its childs. So the interaction between camera render order and render bins order looks like "for each camera { for each render bin }".

Important: the nested cameras (i.e. in the scene graph) render order do not count in the overall camera sorting, only in the currently nesting camera. That is, given a "POST_RENDER 2" main camera, a camera PRE_RENDER will be rendered before, and a PRE_RENDER nested camera will be rendered after, even if it has a smaller render order num.
